apps/lasp/src/lasp_core.erl

```diff
@@ -339,15 +346,31 @@ bind(Id, Value, Store) ->
 -spec bind(id(), value(), function(), store()) -> {ok, var()}.
 bind(Id, {delta, Value}, MetadataFun, Store) ->
     Mutator = fun(#dv{type=Type, metadata=Metadata0, value=Value0,
-                      waiting_threads=WT}=Object) ->
+                      waiting_threads=WT, delta_counter=Counter0,
+                      delta_map=DeltaMap0, delta_ack_map=AckMap}=Object) ->
             Metadata = MetadataFun(Metadata0),
             %% Merge may throw for invalid types.
             try
                 Merged = lasp_type:merge(Type, Value0, Value),
                 {ok, SW} = reply_to_all(WT, [], {ok, {Id, Type, Metadata, Merged}}),
-                NewObject = #dv{type=Type, metadata=Metadata, value=Merged, waiting_threads=SW},
-                %% Return value is a delta state
-                {NewObject, {ok, {Id, Type, Metadata, {delta, Value, Merged}}}}
+                OldValue = lasp_type:query(Type, Value0),
```

Comment:
On Tue, 08 Mar 2016 14:32:55 -0800 "Christopher S. Meiklejohn"
<notifications@github.com> wrote

> >              Metadata = MetadataFun(Metadata0),
> >              %% Merge may throw for invalid types.
> >              try
> >                  Merged = lasp_type:merge(Type, Value0, Value),
> >                  {ok, SW} = reply_to_all(WT, [], {ok, {Id, Type, Metadata,
> >                  Merged}}), > -                NewObject = #dv{type=Type,
> >                  metadata=Metadata, value=Merged, waiting_threads=SW}, > - 
> >                                %% Return value is a delta state > -        
> >                                       {NewObject, {ok, {Id, Type, Metadata,
{delta, Value, Merged}}}}
> > +                OldValue = lasp_type:query(Type, Value0),
> 
> How is this safe to propagate?  'query' returns the actual value and not the
> CRDT structure.  For a G-Counter, this is going to be just an integer; same
> for a PN-counter, so 'NewValue' could equal 'OldValue', but they could be the
> results of completely different options, no? 

I was looking for the way of checking whether deltas are already merged, and I
think that if the old value (X) and the merged value (X join d) are equal, it
means that the deltas were merged to the old value before. I thought that I
should compare with actual values not crdt structures. If there is a better way
to compare them, I will use it. What do you think?

>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/lasp-lang/lasp/pull/81/files#r55440488